## Dividir dataset en 2 conjuntos para entrenamiento y testing

La devisión de un dataset de datos se hace previo al proceso de analisís del modelo, se usa la distribución 80/20 para verificar si el modelo usado es efectivo para predecir variables.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as mlp

In [4]:
data=pd.read_csv("/Users/milan/OneDrive/Documentos/python-ml-course-master/datasets/ads/Advertising.csv")

In [10]:
a=np.random.randn(len(data))
check= (a<0.8)
training=data[check]
testing=data[~check]
len(training), len(testing)

(165, 35)

In [11]:
import statsmodels.formula.api as smf

In [13]:
lm= smf.ols(formula="Sales~TV+Radio", data=training).fit()
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.896
Model:                            OLS   Adj. R-squared:                  0.895
Method:                 Least Squares   F-statistic:                     700.6
Date:                Tue, 18 Aug 2020   Prob (F-statistic):           1.81e-80
Time:                        20:29:58   Log-Likelihood:                -320.70
No. Observations:                 165   AIC:                             647.4
Df Residuals:                     162   BIC:                             656.7
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.5610      0.342      7.486      0.000       1.885       3.237
TV             0.0473      0.002     30.235      0.000       0.044       0.050
Radio          0.1916      0.009     21.699      0.000       0.174       0.209
==============================================================================
Omnibus:                       52.411   Durbin-Watson:                   2.046
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              130.415
Skew:                          -1.341   Prob(JB):                     4.79e-29
Kurtosis:                       6.432   Cond. No.                         446.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Teniendo el modelo y sabiendo los coeficientes del modelo procedemos a usarl en conjunto de testing para validar el modelo.**

In [24]:

sales_pred=lm.predict(testing)
SSD=sum((testing["Sales"]-sales_pred)**2)
RSE = np.sqrt(SSD/(len(testing)-2-1))
ProbError = RSE/np.mean(testing["Sales"])

RSE, ProbError

(1.6809870938288272, 0.13132711670537714)

** Se determina que el error estandar residual es 1.6809870938288272, lo que equivale a un error del 13.13% **